url_shortener/cmd/httpload/main.go

```diff
@@ -0,0 +1,291 @@
+`
+)
+
+type results []result
+
+type result struct {
+	status int
+	timing int64
+}
+
+var (
+	nWorkers    = flag.Int("w", 50, "number of concurrent workers running (default: 50)")
+	nRequests   = flag.Int("n", 200, "number of requests to run (default: 200)")
+	appDuration = flag.Duration("z", 0, "application duration to send requests (default: unlimited)")
+)
+
+func dumpStatuses(results results) string {
+	dumpBuilder := &strings.Builder{}
+
+	statuses := make(map[int]int)
+
+	for _, result := range results {
+		statuses[result.status]++
+	}
+
+	keys := make([]int, 0, len(statuses))
+
+	for key := range statuses {
+		keys = append(keys, key)
+	}
+
+	sort.Ints(keys)
+
+	for _, key := range keys {
+		code := float64(key)
+		count := float64(statuses[key])
+		fmt.Fprintf(dumpBuilder, "%s\n", fmt.Sprintf("[%3.f] %12.f response(s)", code, count))
+	}
+
+	return dumpBuilder.String()
+}
+
+func dumpTimings(results results) string {
+	dumpBuilder := &strings.Builder{}
+
+	totalTiming := float64(0)
+	slowestTiming := float64(0)
+	fastestTiming := float64(0)
+	averageTiming := float64(0)
+	requestsPerSecond := float64(0)
+
+	n := len(results)
+
+	if n > 0 {
+		timing0 := float64(results[0].timing) * nanoSecondRatio
+		totalTiming = timing0
+		slowestTiming = timing0
+		fastestTiming = timing0
+
+		for i := 1; i < n; i++ {
+			timing := float64(results[i].timing) * nanoSecondRatio
+
+			totalTiming += timing
+
+			// the following conditions are counter-intuitive,
+			// but a bigger timing means a slower execution and
+			// a tinier timing means a faster execution
+			// this is the reason I prefer to measure speed
+			// instead of time :)
+			if timing > slowestTiming {
+				slowestTiming = timing
+			}
+
+			if timing < fastestTiming {
+				fastestTiming = timing
+			}
+		}
+
+		averageTiming = totalTiming / float64(n)
+		requestsPerSecond = float64(1) / averageTiming
+	}
+
+	fmt.Fprintf(dumpBuilder, "%s\n", fmt.Sprintf("Total:        %12.4f secs", totalTiming))
+	fmt.Fprintf(dumpBuilder, "%s\n", fmt.Sprintf("Slowest:      %12.4f secs", slowestTiming))
+	fmt.Fprintf(dumpBuilder, "%s\n", fmt.Sprintf("Fastest:      %12.4f secs", fastestTiming))
+	fmt.Fprintf(dumpBuilder, "%s\n", fmt.Sprintf("Average:      %12.4f secs", averageTiming))
+	fmt.Fprintf(dumpBuilder, "%s\n", fmt.Sprintf("Requests/sec: %12.4f", requestsPerSecond))
+
+	return dumpBuilder.String()
+}
+
+func checkFlags(nWorkers int, nRequests int, appDuration time.Duration) (err error) {
+	if nRequests < nWorkers {
+		err = fmt.Errorf("the number of requests to run (%v) cannot be less than the number of workers (%v)", nRequests, nWorkers)
+	}
+	return err
+}
+
+func httpDo(ctx context.Context, req *http.Request, f func(*http.Response, error) error) error {
+	c := make(chan error, 1)
+	req = req.WithContext(ctx)
+	go func() { c <- f(http.DefaultClient.Do(req)) }()
+	select {
+	case <-ctx.Done():
+		<-c // Wait for f to return.
+		return ctx.Err()
+	case err := <-c:
+		return err
+	}
```

Comment:
If the context you pass to the request (as you do in `WithContext`) gets cancelled before the end of the request, the request will fail (`Do() returns an error`).
Now, to know if the request failed because of the context you have to analyze the error returned by Do.
See an example here: https://play.golang.org/p/RI9TTyvz0xO

Calling `ctx.Done()` yourself is fragile for that purpose, since the context might expire between the call to `DefaultClient.Do` and the `<-ctx.Done`

Reply:
I followed the documentation at: https://blog.golang.org/context/google/google.go

What code do you suggest to write?